### ISSUE TITLE
Lua for wavetables

### ISSUE THREAD

**Issue by baconpaul** [surge-synthesizer/surge#4539]
So if you give me

s (power of two samples)
n (number of tables)
f(n) -> float[s] (function to generate table n)

I can easily make a wavetable

and that sounds like a pretty easy thing to do with our lua evaluator

slightly different ui and storage strategy that formula modulator but most of what we have will carry over (which is why i named the gui classes LuaBlah not FormulaBlah).

**Comment by mkruselj** [comment]
For inspiration in what sort of functionality to add:

[Hive Wavetables.zip](https://github.com/surge-synthesizer/surge/files/6480916/Hive.Wavetables.zip)

**Comment by SpotlightKid** [comment]
The Waldorf Microwave (I/II) also had an interesting way of generating Wavetables on the fly. You had to write a Wavtable program in a assembly-like language and then convert that into a Sysex-File, which you then sent to the synth.

I wrote a Python re-implementation of that converter ages ago. The actual calculation of the wavetable program happens in the hardware, though:

https://github.com/SpotlightKid/upaw2/blob/master/upaw.description

**Comment by baconpaul** [comment]
Gonna bounce this out of 1.2 again and deactivate it in the nightly

**Comment by nuoun** [comment]
So the wavetable script editor has been re-enabled in the Nightly builds for a while and I've been slowly working on it, so far #7674, #7680, #7687, #7755, #7764 and #7767 have been merged which refactor the code, make UI changes, add window tearout, implement the prelude for WTSE and some more and things are slowly coming together. 
I had a much longer writeup ready which I could still post if there is interest in some of the concepts and ideas and possible side projects for the future but I'll try to keep this short and limit this post to priority here which is to try and figure out what is needed to get this component ready for release.

So I will add a (likely incomplete) list of issues that still need work in the next post, but a lot of this is to just test and periodically ask if stuff makes sense or what needs improving and I would of course much appreciate feedback on this. Paul brought up two things on a previous PR which I'll mention here:

1. Window tear out, this currently works different for the WTSE, in formula the editor switches to the next LFO when a different LFO that has formula active gets selected and the WTSE does not do this even when another WT oscillator gets selected. I don't think this is a big issue, maybe a QoL thing at best and I'm not sure how this is handled in the GUI code right now. Other than that, the torn out window just stays open when the VCO gets changed and I think that is in line with how other torn out windows behave.

2. Accessibility, this needs both improvements and testing, the main issue I see now is that the graph's current frame numberfield is probably not helpful since accessibility should be disable on that entire component anyway so it should also skip over that UI element. Then it needs more testing but it probably makes sense to do that after we decide if this layout is final and/or if there is more to add.

So TL;DR, this project is being worked on and feel free to discuss, suggest, critique, help test, get involved, etc. (But really Shortcircuit could use more help so if you're looking for a project please focus on that!)

**Comment by nuoun** [comment]
Outstanding issues (WIP):

Code
- Add formula style init function
- Bail out of generate loop on Lua error
- Cancel button during generate
- Rename Lua variables and rename config table to wt in default script
- Add CI test for WTSE Prelude

UI
- Make mouse drag on the graph adjust sensitivity based on wavetable size
- Add menu links to manual on UI elements
- Discuss and ensure proper torn out window functionality

WTSE Prelude
- Add table versions of envelope, slew limiter, a simple compressor, etc
- Fourier series coefficients for common waveshapes
- More (numpy/matlab inspired) math functions

Accessibility
- Ensure proper naming and correct tab order of buttons
- Disable accessibility navigation on graph button
- Testing

Documentation 
- Manual section

Tutorial and examples
- Basic tutorial that goes over wavetable concepts
- Tutorial that demonstrates various concepts like windowing, normalization, envelope, etc
- Additive synthesis examples with modulation over frames (e.g. types of FM/AM)
- PolyBLEP examples with modulation (e.g. PWM, phase modulation, phase distortion)
- Some other bandlimited synthesis methods (e.g. padsynth, BP resonator)

Bugfixes
- Fix graph not following onSkinChanged()
- More testing!

**Comment by nuoun** [comment]
Few suggestions/issues that came up recently

WTSE overlay
    - WTSE editor should refresh on patch load
    - WTSE preview should remember last used view mode on opening

Lua sandbox
    - error(nil) causes crash
    - print(nil) should do nothing rather than throw error
    
Minor issues
    - Wavetable file export should not write special chars in filenames, maybe replace these with _
    - Default WTSE script clips slightly
    - Copying WT oscillator should also copy the current WTSE script when copying to another oscillator

**Comment by nuoun** [comment]
Found another stack leak that happens on certain syntax errors, a faulty assignment, unexpected use of an operators or anything that causes to generate function to break results in
```
Surge Error [Wavetable Script Evaluator]
Unable to locate generate function
Guarded stack leak: [WavetableScript::evaluateInner] exit=6 enter=5
  6 -> nil
  5 -> nil
```

To reproduce do something like these lines
```
test = -- faulty assignment
test = 1 * -- missing operand
thisstatementisincomplete
```

**Comment by baconpaul** [comment]
These occur when an error path with a return is missing a pop inside the scope of the guard

**Comment by mkruselj** [comment]
I think we can close this issue, we have the feature working, the rest is just expanding on its possibilities as we go along.

**Comment by nuoun** [comment]
I think most of what I listed above was fixed (need to test) but there's still some outstanding issues:

- Script error in filmstrip mode repeatedly throws the same error
- Addition of the filmstrip mode introduced some sort of caching where a WT only re-generates with actual script changes which is not great for scripts that rely on math.random()
